From f3cc54c3c3d4e74105124c464b0e93caa072681f Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:02:58 -0400
Subject: [fs] btrfs: fix use-after-free in mount_subvol()

Message-id: <1407179109-13714-292-git-send-email-sandeen@redhat.com>
Patchwork-id: 87213
O-Subject: [RHEL7.1 PATCH 291/422] btrfs: fix use-after-free in mount_subvol()
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Christoph Jaeger <christophjaeger@linux.com>

commit 0040e606e35a0db80fc3fac04ccc7c7176a8e2b1
Author: Christoph Jaeger <christophjaeger@linux.com>
Date:   Sat Apr 12 13:33:13 2014 +0200

    btrfs: fix use-after-free in mount_subvol()

    Pointer 'newargs' is used after the memory that it points to has already
    been freed.

    Picked up by Coverity - CID 1201425.

    Fixes: 0723a0473f ("btrfs: allow mounting btrfs subvolumes with
    different ro/rw options")
    Signed-off-by: Christoph Jaeger <christophjaeger@linux.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 994c409..53bc373 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -1186,7 +1186,6 @@ static struct dentry *mount_subvol(const char *subvol_name, int flags,
   return ERR_PTR(-ENOMEM);
  mnt = vfs_kern_mount(&btrfs_fs_type, flags, device_name,
         newargs);
- kfree(newargs);
 
  if (PTR_RET(mnt) == -EBUSY) {
   if (flags & MS_RDONLY) {
@@ -1196,17 +1195,22 @@ static struct dentry *mount_subvol(const char *subvol_name, int flags,
    int r;
    mnt = vfs_kern_mount(&btrfs_fs_type, flags | MS_RDONLY, device_name,
           newargs);
-   if (IS_ERR(mnt))
+   if (IS_ERR(mnt)) {
+    kfree(newargs);
     return ERR_CAST(mnt);
+   }
 
    r = btrfs_remount(mnt->mnt_sb, &flags, NULL);
    if (r < 0) {
     /* FIXME: release vfsmount mnt ??*/
+    kfree(newargs);
     return ERR_PTR(r);
    }
   }
  }
 
+ kfree(newargs);
+
  if (IS_ERR(mnt))
   return ERR_CAST(mnt);
 
-- 
1.7.1